File: fluttermultiplatform/lib/models/product.dart
import 'user.dart';
import 'batch.dart';
import 'product_image.dart';
import 'review.dart';

class Product {
  final String id;
  final String userId;
  final String? batchId;
  final String name;
  final String description;
  final double price;
  final int stock;
  final String shrimpType;
  final String size;
  final String unit;
  final bool isAvailable;
  final double ratingAvg;
  final int ratingCount;
  final User? user;
  final Batch? batch;
  final List<ProductImage>? images;
  final List<Review>? reviews;
  final String createdAt;

  Product({required this.id, this.userId = '', this.batchId, required this.name, this.description = '', this.price = 0, this.stock = 0, this.shrimpType = '', this.size = '', this.unit = 'kg', this.isAvailable = true, this.ratingAvg = 0, this.ratingCount = 0, this.user, this.batch, this.images, this.reviews, this.createdAt = ''});

  factory Product.fromJson(Map<String, dynamic> json) => Product(
        id: json['id']?.toString() ?? '',
        userId: json['user_id']?.toString() ?? '',
        batchId: json['batch_id']?.toString(),
        name: json['name'] ?? '',
        description: json['description'] ?? '',
        price: (json['price'] ?? 0).toDouble(),
        stock: (json['stock'] ?? 0).toInt(),
        shrimpType: json['shrimp_type'] ?? '',
        size: json['size'] ?? '',
        unit: json['unit'] ?? 'kg',
        isAvailable: json['is_available'] ?? true,
        ratingAvg: (json['rating_avg'] ?? 0).toDouble(),
        ratingCount: (json['rating_count'] ?? 0).toInt(),
        user: json['user'] != null ? User.fromJson(json['user']) : null,
        batch: json['batch'] != null ? Batch.fromJson(json['batch']) : null,
        images: json['images'] != null ? (json['images'] as List).map((e) => ProductImage.fromJson(e)).toList() : null,
        reviews: json['reviews'] != null ? (json['reviews'] as List).map((e) => Review.fromJson(e)).toList() : null,
        createdAt: json['created_at'] ?? '',
      );

  Map<String, dynamic> toJson() => {
        'id': id, 'name': name, 'description': description, 'price': price, 'stock': stock, 'shrimp_type': shrimpType, 'size': size, 'unit': unit, 'is_available': isAvailable, 'rating_avg': ratingAvg, 'rating_count': ratingCount,
      };
}

File: backend/models/product.go
package models
import (
	"time"
	"github.com/google/uuid"
	"gorm.io/gorm"
)
type Product struct {
	ID          uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	UserID      uuid.UUID      `gorm:"type:uuid;not null;index" json:"user_id"`
	BatchID     *uuid.UUID     `gorm:"type:uuid;index" json:"batch_id"`
	Name        string         `gorm:"type:varchar(255);not null" json:"name" validate:"required,min=2,max=255"`
	Description string         `gorm:"type:text" json:"description"`
	Price       float64        `gorm:"type:decimal(15,2);not null" json:"price" validate:"required,gt=0"`
	Stock       int            `gorm:"not null;default:0" json:"stock"`
	ShrimpType  string         `gorm:"type:varchar(100)" json:"shrimp_type"`
	Size        string         `gorm:"type:varchar(50)" json:"size"`
	Unit        string         `gorm:"type:varchar(20);default:'kg'" json:"unit"`
	IsAvailable bool           `gorm:"default:true" json:"is_available"`
	RatingAvg   float64        `gorm:"type:decimal(3,2);default:0" json:"rating_avg"`
	RatingCount int            `gorm:"default:0" json:"rating_count"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
	User    User           `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Batch   *Batch         `gorm:"foreignKey:BatchID" json:"batch,omitempty"`
	Images  []ProductImage `gorm:"foreignKey:ProductID" json:"images,omitempty"`
	Reviews []Review       `gorm:"foreignKey:ProductID" json:"reviews,omitempty"`
}
func (p *Product) BeforeCreate(tx *gorm.DB) error {
	if p.ID == uuid.Nil {
		p.ID = uuid.New()
	}
	return nil
}
type ProductImage struct {
	ID        uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	ProductID uuid.UUID      `gorm:"type:uuid;not null;index" json:"product_id"`
	ImageURL  string         `gorm:"type:varchar(500);not null" json:"image_url"`
	IsPrimary bool           `gorm:"default:false" json:"is_primary"`
	CreatedAt time.Time      `json:"created_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}
func (pi *ProductImage) BeforeCreate(tx *gorm.DB) error {
	if pi.ID == uuid.Nil {
		pi.ID = uuid.New()
	}
	return nil
}
